### ISSUE TITLE
Facing issues with setclip and setcroppingpolygons

### ISSUE THREAD

**Issue by prasadrc** [openseadragon/openseadragon#2375]
Hi

I would like to know, can we use both setclip and setcroppingpolygon at once.
The requirement is, an image will be clipped and displayed to the user, after that user will select certain region using mouse and those coordinates need to be cropped and displayed to the user.kindly suggest appropriate approach.

**Comment by prasadrc** [comment]
@iangilman  kindly provide suggestion.

**Comment by pearcetm** [comment]
I haven’t tried using cropping polygons and a clip rectangle at the same time. Have you tried it? What happens?

**Comment by iangilman** [comment]
Indeed... I don't know if they both work together! Looks like they might... Did you give it a try?

**Comment by prasadrc** [comment]
Yes I tried.It displays, but not as expected.when I try to zoom in zoom out it is not working as expected.So any other solution to achieve the requirement.

**Comment by iangilman** [comment]
Interesting. I don't know if anyone has tested this before, but it certainly sounds like a bug.

Meanwhile, if you want to work around it, you should be able to incorporate the clipping rectangle into the cropping polygons by including that rectangle as part of the polygons.

**Comment by prasadrc** [comment]
Hi do we have any other ways to achieve desired requirement.

**Comment by pearcetm** [comment]
I discovered what the problem is, and there's a relatively easy fix. When clipping and cropping polygons are both used, the context state is saved twice and only restored once. The fix is to only save the context state the first time, so it is restored properly after drawing.

https://github.com/openseadragon/openseadragon/blob/ffbd8f985a4dab4fdde855458383f50f6973eb7a/src/tiledimage.js#L2045-L2047

Should be:
```
        if (this._croppingPolygons) {
            var self = this;
            if(!usedClip){
                this._drawer.saveContext(useSketch);
            }
```

I'm implementing this in the canvas drawer in https://github.com/openseadragon/openseadragon/pull/2310

**Comment by prasadrc** [comment]
Thank you.Once the issue is fixed kindly let me know.

**Comment by iangilman** [comment]
@pearcetm Bravo on the fix! 

@prasadrc The PR @pearcetm put his fix in (#2310) is a very large new feature that probably won't be ready for some time. If you'd like the fix sooner, you should follow his instructions for how to make the change yourself.

**Comment by prasadrc** [comment]
var usedClip = false;
        if ( this._clip ) {
            this._drawer.saveContext(useSketch);

            var box = this.imageToViewportRectangle(this._clip, true);
            box = box.rotate(-this.getRotation(true), this._getRotationPoint(true));
            var clipRect = this._drawer.viewportToDrawerRectangle(box);
            if (sketchScale) {
                clipRect = clipRect.times(sketchScale);
            }
            if (sketchTranslate) {
                clipRect = clipRect.translate(sketchTranslate);
            }
            this._drawer.setClip(clipRect, useSketch);

            usedClip = true;
        }

        if (this._croppingPolygons) {
            var self = this;
            this._drawer.saveContext(useSketch);
            try {
                var polygons = this._croppingPolygons.map(function (polygon) {
                    return polygon.map(function (coord) {
                        var point = self
                            .imageToViewportCoordinates(coord.x, coord.y, true)
                            .rotate(-self.getRotation(true), self._getRotationPoint(true));
                        var clipPoint = self._drawer.viewportCoordToDrawerCoord(point);
                        if (sketchScale) {
                            clipPoint = clipPoint.times(sketchScale);
                        }
                        if (sketchTranslate) {
                            clipPoint = clipPoint.plus(sketchTranslate);
                        }
                        return clipPoint;
                    });
                });
                this._drawer.clipWithPolygons(polygons, useSketch);
            } catch (e) {
                $.console.error(e);
            }
            usedClip = true
Kindly excuse,I think  code is written as expected.

**Comment by prasadrc** [comment]
Thankyou.I understand what is the issues is and It worked 🎉😊.
